airbyte-cdk/bulk/toolkits/load-object-storage/src/main/kotlin/io/airbyte/cdk/load/file/object_storage/PartFactory.kt

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.cdk.load.file.object_storage
+
+import java.util.concurrent.ConcurrentLinkedQueue
+import java.util.concurrent.atomic.AtomicReference
+
+/**
+ * Generates part w/ metadata for a multi-part upload for a given key and file no. parts are
+ * 1-indexed. For convenience, empty parts are tolerated but not counted by the assembler.
+ *
+ * Not thread-safe. It is expected that the parts are generated in order.
+ */
+class PartFactory(
+    val key: String,
+    val fileNumber: Long,
+) {
+    var totalSize: Long = 0
+    private var nextIndex: Int = 0
+    private var finalProduced = false
+
+    fun nextPart(bytes: ByteArray?, isFinal: Boolean = false): Part {
```

Comment:
Under what circumstances would bytes be null and why do we want to call this method? Is it just for flushing at the end?

Reply:
Partly it's for convenience: if there's no new data everything still just works. You produce an empty part and the assembler ignores it.

But also it's normal for the final part to sometimes be empty -- for example if all you get in the part is an EOS. In that case `isFinal` is true and `bytes` is `null`. Int his case the assembler treats this as "fact of completion" and infers the number of parts from the index, but doesn't add a new part.

I'll add comments clarifying this.

Reply:
I'd think if the final part is null it isn't a part per se. It feels like we're conflating upload finalization with part loading. Not the worst, but we have to add `!= null` branching to every downstream layer which makes things harder to follow. A single hard branch would improve readability and testability imho.